tests/tdastro/sources/test_snia.py

```diff
@@ -0,0 +1,174 @@
+    times = t0 + phase_obs
+
+    if opsim:
+        ra = source.get_param(state, "ra")
+        dec = source.get_param(state, "dec")
+        times = opsim_data.get_observed_times(ra, dec, radius=1.75)
+        phase_obs = times - t0
+        times = np.sort(times[(phase_obs > -20) & (phase_obs < 50)])
+        # Note that we don't have filter info yet.
+
+        if len(times) == 0:
+            print(f"No overlap time in opsim for (ra,dec)=({ra:.2f},{dec:.2f})")
+            return
+
+    res["times"] = times
+
+    flux_flam = source.evaluate(times, wave_obs, graph_state=state)
+    res["flux_flam"] = flux_flam
+
+    # convert ergs/s/cm^2/AA to nJy
+
+    flux_fnu = flam_to_fnu(
+        flux_flam, wave_obs, wave_unit=u.AA, flam_unit=u.erg / u.second / u.cm**2 / u.AA, fnu_unit=u.nJy
+    )
+
+    res["flux_fnu"] = flux_fnu
+
+    bandfluxes = passbands.fluxes_to_bandfluxes(flux_fnu, wave_obs)
+    res["bandfluxes"] = bandfluxes
+
+    res["state"] = state
+
+    return res
+
+
+def test_snia_end2end(opsim=False, nsample=1, return_result=False, phase_rest=None, wavelengths_rest=None):
+    """Test that we can sample and create SN Ia simulation using the salt3 model."""
+
+    # Create a host galaxy anywhere on the sky.
+    host = SNIaHost(
+        ra=NumpyRandomFunc("uniform", low=0.0, high=360.0),
+        dec=NumpyRandomFunc("uniform", low=-90.0, high=33.5),
+        hostmass=NumpyRandomFunc("uniform", low=7, high=12),
+        redshift=NumpyRandomFunc("uniform", low=0.01, high=0.02),
+    )
+
+    distmod_func = DistModFromRedshift(host.redshift, H0=73.0, Omega_m=0.3)
+    x1_func = HostmassX1Func(host.hostmass)
+    c_func = NumpyRandomFunc("normal", loc=0, scale=0.02)
+    m_abs_func = NumpyRandomFunc("normal", loc=-19.3, scale=0.1)
+
+    x0_func = X0FromDistMod(
+        distmod=distmod_func,
+        x1=x1_func,
+        c=c_func,
+        alpha=0.14,
+        beta=3.1,
+        m_abs=m_abs_func,
+    )
+
+    sncosmo_modelname = "salt2-h17"
+
+    source = SncosmoWrapperModel(
+        sncosmo_modelname,
+        t0=NumpyRandomFunc("uniform", low=60796, high=64448),
+        x0=x0_func,
+        x1=x1_func,
+        c=c_func,
+        ra=NumpyRandomFunc("normal", loc=host.ra, scale=0.01),
+        dec=NumpyRandomFunc("normal", loc=host.dec, scale=0.01),
+        redshift=host.redshift,
+    )
+
+    source.add_effect(Redshift(redshift=source.redshift, t0=source.t0))
+
+    passbands = PassbandGroup(
+        passbands=[
+            Passband(
+                "LSST", "g", table_url="https://github.com/lsst/throughputs/blob/main/baseline/total_g.dat"
+            ),
+            Passband(
+                "LSST", "r", table_url="https://github.com/lsst/throughputs/blob/main/baseline/total_r.dat"
+            ),
+        ]
+    )
+
+    if opsim:
+        opsim_file = "/Users/mi/Work/tdastro/opsim_db/baseline_v3.4_10yrs.db"
+        opsim_table = OpSim.from_db(opsim_file)
+        opsim_data = OpSim(opsim_table)
```

Comment:
Should we remove it / change to downloading file from the Internet?

Reply:
I don't think we need the opsim for this particular test.

Reply:
I'd like to have it for an end-to-end test, but the opsim db file is too large to download or load. We should probably put in a lighter version of the opsim file for testing?